devolutions-agent/src/updater/io.rs

```diff
@@ -0,0 +1,76 @@
+//! IO utilities for the updater logic
+
+use camino::Utf8Path;
+use camino::Utf8PathBuf;
+use futures::TryFutureExt;
+use tokio::{fs::File, io::AsyncWriteExt};
+
+use crate::updater::UpdaterError;
+
+/// Download binary file to memory
+pub async fn download_bytes(url: &str) -> Result<Vec<u8>, UpdaterError> {
+    info!(%url, "Downloading file from network...");
+
+    let body = reqwest::get(url)
+        .and_then(|response| response.bytes())
+        .map_err(|source| UpdaterError::FileDownload {
+            source,
+            file_path: url.to_string(),
+        })
+        .await?;
+    Ok(body.to_vec())
+}
+
+/// Download UTF-8 file to memory
+pub async fn download_text(url: &str) -> Result<String, UpdaterError> {
+    let bytes = download_bytes(url).await?;
+    String::from_utf8(bytes).map_err(|_| UpdaterError::Utf8)
+}
+
+/// Save data to a temporary file
+pub async fn save_to_temp_file(data: &[u8], extension: Option<&str>) -> Result<Utf8PathBuf, UpdaterError> {
+    let uuid = uuid::Uuid::new_v4();
+
+    let file_name = match extension {
+        Some(ext) => format!("{uuid}.{}", ext),
+        None => uuid.to_string(),
+    };
+
+    let file_path = Utf8PathBuf::from_path_buf(std::env::temp_dir())
+        .expect("BUG: OS Should always return valid UTF-8 temp path")
+        .join(file_name);
+
+    let mut file = File::create(&file_path).await?;
+    file.write_all(data).await?;
+
+    remove_file_on_reboot(&file_path)?;
+
+    Ok(file_path)
+}
+
+/// Mark file to be removed on next reboot.
+pub fn remove_file_on_reboot(file_path: &Utf8Path) -> Result<(), UpdaterError> {
+    _impl_remove_file_on_reboot(file_path)
+}
+
+#[cfg(windows)]
+pub fn _impl_remove_file_on_reboot(file_path: &Utf8Path) -> Result<(), UpdaterError> {
```

Comment:
naming: Do not use `_` in prefix unless you want to suppress the "unused" / "dead_code" warnings.
Also, I think we mostly put `impl` as a suffix in the rest of the codebase.
```suggestion
pub fn remove_file_on_reboot_impl(file_path: &Utf8Path) -> Result<(), UpdaterError> {
```